lib/iris/_shapefiles.py

```diff
@@ -0,0 +1,221 @@
+    if not target_system:
+        warnings.warn(
+            "Cube has no coord_system; using default GeogCS lat/lon",
+            IrisUserWarning,
+        )
+        target_system = DEFAULT_CS
+    if geometry_system is None:
+        geometry_system = DEFAULT_CS
+    target_proj = target_system.as_cartopy_projection()
+    source_crs = geometry_system.as_cartopy_projection()
+
+    trans_geometry = target_proj.project_geometry(geometry, source_crs)
+    if target_system == DEFAULT_CS and cube.coord(x_name).points[-1] > 180:
+        trans_geometry = shapely.transform(trans_geometry, _trans_func)
+
+    return trans_geometry
+
+
+def _trans_func(geometry):
+    """pocket function for transforming the x coord of a geometry from -180 to 180 to 0-360"""
+    for point in geometry:
+        if point[0] < 0:
+            point[0] = 360 - np.abs(point[0])
+    return geometry
+
+
+def _get_global_cs():
+    """pocket function for returning the iris default coord system to avoid circular imports"""
+    import iris.fileformats.pp
+
+    DEFAULT_CS = iris.coord_systems.GeogCS(iris.fileformats.pp.EARTH_RADIUS)
+    return DEFAULT_CS
+
+
+def _cube_primary_xy_coord_names(cube):
+    """Return the primary latitude and longitude coordinate standard names, or
+    long names, from a cube.
+
+    Arguments:
+        cube (:class:`iris.cube.Cube`): An Iris cube
+
+    Returns:
+        The names of the primary latitude and longitude coordinates
+    """
+    latc = cube.coords(axis="y")[0] if cube.coords(axis="y") else -1
+    lonc = cube.coords(axis="x")[0] if cube.coords(axis="x") else -1
+
+    if -1 in (latc, lonc):
+        msg = "Error retrieving xy dimensions in cube: {!r}"
+        raise ValueError(msg.format(cube))
+
+    latitude = latc.standard_name if latc.standard_name else latc.long_name
+    longitude = lonc.standard_name if lonc.standard_name else lonc.long_name
+    return latitude, longitude
+
+
+def _cube_xy_guessbounds(cube):
+    """Guess latitude/longitude bounds of the cube and add them (**in place**)
+    if not present.
+
+    Arguments:
+        cube (:class:`iris.cube.Cube`): An Iris cube
+
+    Warning:
+        This function modifies the passed `cube` in place, adding bounds to the
+        latitude and longitude coordinates.
+    """
+    for coord in _cube_primary_xy_coord_names(cube):
+        if not cube.coord(coord).has_bounds():
+            cube.coord(coord).guess_bounds()
+
+
+def _rebase_values_to_modulus(values, modulus):
+    """Rebase values to a modulus value.
+
+    Arguments:
+        values (list, tuple): The values to be re-based
+        modulus             : The value to re-base to
+
+    Returns:
+        A list of re-based values.
+    """
+    rebased = []
+    for val in values:
+        nv = np.abs(val) % modulus
+        if val < 0.0:
+            nv *= -1.0
+        if np.abs(nv - modulus) < 1e-10:
+            nv = 0.0
+        rebased.append(nv)
+    return rebased
```

Comment:
(I'm working on an idea for this at the moment)

Reply:
After implementing this idea: https://github.com/SciTools/iris/pull/5470#discussion_r1400698757, the below idea shaves off another 50%:

```python
def _get_mod_rebased_coord_bounds(coord):
    modulus = coord.units.modulus
    # Force realisation (rather than core_bounds) - more efficient for the
    #  repeated indexing happening downstream.
    result = coord.bounds
    if modulus:
        result = np.abs(result) % modulus
        result[result < 0.0] *= -1
        result[np.isclose(result, modulus, 1e-10)] = 0.0

    return result
```

You generate the rebased bounds once - before the loop - then just index that new array within the loop. I believe it should produce the same results.

Reply:
This function doesn't work the same as the previous version because the negative result values are returned as positive.  For example, You need to do abs(-387) & 360 = 27, then see the original -387 was negative and convert 27 to -27 (because -387 % 360 is 333). This function converts everything to postive values, and as such the second line never gets called because it's empty. 

I tried replacing with
`def _get_mod_rebased_coord_bounds(coord):
    modulus = coord.units.modulus
    # Force realisation (rather than core_bounds) - more efficient for the
    #  repeated indexing happening downstream.
    result = coord.bounds
    if modulus:
        result[result > 0.0] = result % modulus
        result[result < 0.0] = (np.abs(result) % modulus)*-1
        result[np.isclose(result, modulus, 1e-10)] = 0.0`

But get `ValueError: assignment destination is read-only`

Reply:
Thanks for catching that!
The `ValueError` is also my mistake - I was clearly testing with something non-representative of the real world. We don't want to edit `coord.bounds` - we actually want a new copy.

```python
result = np.array(coord.bounds)
```